Hi,
I don't seem to be able to get flambda to install. Is there a canonical way
to do it?

$ opam config env --switch  4.05.0+flambda
[ERROR] The current switch (4.05.0+flambda) is an unknown compiler switch.
$ sudo apt-get install ocaml
ocaml is already the newest version (4.02.3-5ubuntu2).
$ opam install base-flambda
base-flambda is not available because your system doesn't comply with ..
<all versions printed are above 4.02.3>

Is flambda only available in a non-distributed version of ocaml? If so how
to I get it? Do I build from source? How?

Hope you can help
Thanks,
Anton


On Wed, 11 Oct 2017 at 18:46 Simon Cruanes <notifications@github.com> wrote:

> *@c-cube* requested changes on this pull request.
>
> That's nice, I will merge, but there are some small things to change.
>
> For OCaml, which version do you use? It would be best to use
> 4.04.0+flambda or 4.05.0+flambda with the -O3 flag put back.
> ------------------------------
>
> In Makefile
> <https://github.com/c-cube/hashset_benchs/pull/2#discussion_r144083902>:
>
> >
> -ocaml:
> -	ocamlfind ocamlopt -O3 neighbors2.ml -o neighbors2_ocaml
>
> why do you remove this flag?
> ------------------------------
>
> In README.md
> <https://github.com/c-cube/hashset_benchs/pull/2#discussion_r144084812>:
>
> >
> -So far, indeed, on my computer, F# is faster (1.93s, where rust takes 2.46s and OCaml 2.24).
> +Jon Harrop>
>
> That was my remark, not Jon's
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/c-cube/hashset_benchs/pull/2#pullrequestreview-68691533>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AfMtw4ySgssBL8uRENubMqIX58Yuv0BQks5srP7ngaJpZM4P12Ou>
> .
>

--- a/Makefile
+++ b/Makefile
@@ -37,7 +37,7 @@ run:
 	@$(time) ./neighbors2_clang++
 
 ocaml neighbors2_ocaml: neighbors2.ml
-	ocamlfind ocamlopt neighbors2.ml -o neighbors2_ocaml
+	ocamlfind ocamlopt -O3 neighbors2.ml -o neighbors2_ocaml
 
 rust neighbors2_rust: neighbors2.rs
 	rustc -O neighbors2.rs -o neighbors2_rust
@@ -52,3 +52,15 @@ clang++ neighbors2_clang++: neighbors2.cc
 	clang++-4.0 --std=c++11 -O3 -o neighbors2_clang++ neighbors2.cc
 
 # ocamlfind ocamlopt -g -O3 -package containers.data -linkpkg neighbors2.ml -o neighbors2_ocaml
+
+# To run ocamlfind ocamlopt -O3 requires flambda
+
+# You will need opam if you don't have it already:
+#   sudo apt install opam
+#   opam update
+#   [see https://opam.ocaml.org/doc/Usage.html#Basics]
+#
+#   opam switch 4.04.0+flambda
+#   opam install ocamlfind
+
+

--- a/README.md
+++ b/README.md
@@ -8,50 +8,10 @@ Trying to reproduce/improve on https://www.reddit.com/r/rust/comments/4dd5yl/rus
 - Clojure
 - C++ (g++ and clang++)
 
-Jon Harrop>
+Simon Cruanes:
 |So far, indeed, on my computer, F# is faster (1.93s, where rust takes 2.46s and OCaml 2.24).
 
-Anton Carver
-Wed 11 Oct 17:23:12 BST 2017
-model name	: Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz
-=== F#
-result is 8000
-real 2.59
-user 2.46
-sys 0.13
-=== OCaml
-result is 8000
-real 4.86
-user 4.86
-sys 0.00
-=== rust
-8000
-real 1.92
-user 1.92
-sys 0.00
-=== clojure
-result is 8000
-real 4.38
-user 8.44
-sys 0.54
-=== clojure_bis
-result is 8000
-real 3.15
-user 6.94
-sys 0.48
-=== C++ (G++)
-8000
-real 2.58
-user 2.58
-sys 0.00
-=== C++ (Clang/LLVM)
-8000
-real 2.60
-user 2.59
-sys 0.00
-
-Order is: Rust (1.92), G++ ~ F# ~ C++ (2.58-2.60), Clojure, OCaml (4.86)
-
+Anton Carver:
 Clojure is presumably hampered by the fact it starts a JVM each time.
 I added an empty Clojure test to measure this:
 
@@ -59,3 +19,49 @@ I added an empty Clojure test to measure this:
 real 1.28
 user 3.97
 sys 0.31
+
+Anton Carver
+Fri 13 Oct 22:06:35 BST 2017
+model name	: Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz
+=== F#
+result is 8000
+real 2.58
+user 2.45
+sys 0.13
+=== OCaml
+result is 8000
+real 2.65
+user 2.65
+sys 0.00
+=== rust
+8000
+real 1.96
+user 1.96
+sys 0.00
+=== clojure
+result is 8000
+real 4.80
+user 9.25
+sys 0.42
+=== clojure_bis
+result is 8000
+real 3.07
+user 6.63
+sys 0.48
+=== clojure_empty (empty program for startup overhead)
+real 1.23
+user 3.91
+sys 0.24
+=== C++ (G++)
+8000
+real 2.54
+user 2.54
+sys 0.00
+=== C++ (Clang/LLVM)
+8000
+real 2.60
+user 2.60
+sys 0.00
+
+Order is: Rust (1.96), G++ ~ F# ~ C++ (2.54-2.60), OCaml (2.65), Clojure
+

--- a/neighbors2.cc
+++ b/neighbors2.cc
@@ -12,7 +12,7 @@ struct PairHash {
 
 typedef std::unordered_set<Pair, PairHash> Set;
 
-bool contains(const Set& s, const Pair& p) {
+inline bool contains(const Set& s, const Pair& p) {
   return s.find(p) != s.end();
 }
 
@@ -30,12 +30,12 @@ const Set nth_loop(int n, Set& s1, Set& s2, Set& s0) {
   if (n == 0) {
     return s1;
   } else {
+    auto add = [&](const Pair& p){
+      if (!(contains(s1, p) || contains(s2, p))) {
+        s0.insert(p);
+      }
+    };
     for (auto p : s1) {
-      auto add = [&](const Pair& p){
-	if (!(contains(s1, p) || contains(s2, p))) {
-	  s0.insert(p);
- 	}
-      };
       iter_neighbors(add, p);
     }
     s2.clear();

